pandas/tests/groupby/test_apply.py

```diff
@@ -1331,3 +1331,19 @@ def test_result_name_when_one_group(name):
     expected = Series([1, 2], name=name)
 
     tm.assert_series_equal(result, expected)
+
+
+def test_empty_df():
+    # GH 47985
+    empty_df = DataFrame({"a": [], "b": []})
+
+    # Both operations should return an empty series instead of IndexError for apply UDF
+    result1 = empty_df.groupby("a", group_keys=True).b.apply(lambda x: x.values[-1])
+    result2 = empty_df.groupby("a", group_keys=True).b.agg("sum")
```

Comment:
Can you parametrize this instead - e.g. have ("apply", lambda x: x.values[-1]) and ("agg", "sum") as the method / op parameters and do

```
gb = empty_df.groupby("a", group_keys=True)
result = getattr(gb, method)(op)
```

This allows developers to debug more easily in the future. The common case you can run into is setting a break point within the code for a test that consists of multiple tests. That break point can then be hit by the successful tests as well as the one that is failing, making debugging more difficult.

By using pytest's parametrize, one can just run the failing test.